pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpsSegmentFetcher.java

```diff
@@ -0,0 +1,169 @@
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509TrustManager;
+import org.apache.commons.configuration.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/*
+ * Note:
+ * The 'server' in this class is not pinot-server.
+ * SegmentFetchers are used in pinot-controller (to fetch segments from external sources),
+ * and pinot-server (to fetch segment from the controller).
+ *
+ * The 'server' here indicates the server side of the HTTPS connection.
+ *
+ * Also note that this is ONE of many combinations of algorithms and types that can be used
+ * for an Https client.
+ *
+ * This implementation will verify the server's certificate against a Certifying Authority (CA)
+ * if X509 certificate of the CA is configured. Otherwise, all server certificates will be treated
+ * as trusted.
+ *
+ * The implementation provides for optional configuration for key managers (where client certificate
+ * is stored). If configured, the client will present its certificate to the server during TLS
+ * protocol exchange.
+ *
+ * At this time, only PKCS12 files are supported for client certificates, and it is assumed that the
+ * server is presenting an X509 certificate.
+ */
+public class HttpsSegmentFetcher implements SegmentFetcher {
+  SSLSocketFactory _sslSocketFactory;
+
+  private static final String SECURITY_ALGORITHM = "TLS";
+  private static final String CERTIFICATE_TYPE = "X509";
+  private static final String KEYSTORE_TYPE = "PKCS12";
+  private static final String KEYMANAGER_FACTORY_ALGORITHM = "SunX509";
+
+  private static String CONFIG_OF_SERVER_CA_CERT = "ssl.server.ca-cert";
+  private static String CONFIG_OF_CLIENT_PKCS12_FILE = "ssl.client.pkcs12.file";
+  private static String CONFIG_OF_CLIENT_PKCS12_PASSWORD = "ssl.client.pkcs12.password";
+
+  public static Logger LOGGER = LoggerFactory.getLogger(HttpsSegmentFetcher.class);
+
+  @Override
+  public void init(Configuration configs) {
+    try {
+      TrustManager[] trustManagers = setupTrustManagers(configs);
+      KeyManager[] keyManagers = setupKeyManagers(configs);
+
+      SSLContext sslContext = SSLContext.getInstance(SECURITY_ALGORITHM);
+      sslContext.init(keyManagers, trustManagers, null);
+      _sslSocketFactory = sslContext.getSocketFactory();
+    } catch (Exception e) {
+      Utils.rethrowException(e);
+    }
+  }
+
+  private TrustManager[] setupTrustManagers(Configuration configs)
+      throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
+    // This is the cert authority that validates server's cert, so we need to put it in our
+    // trustStore.
+    String serverCACertFile = configs.getString(CONFIG_OF_SERVER_CA_CERT);
+    if (serverCACertFile == null) {
```

Comment:
If encryption is top priority, shouldn't we let users enable TrustAllCerts in an explicit flag? I.e., if serverCACertFile is null and useTrustAllCert is true. Otherwise, users might not notice trust manager is accpeting all certificates if s/he is not looking at the log when booting up the machine.

Reply:
Hmm, so u are suggesting adding another config variable. Let me think some more, and maybe we can add one.

Reply:
We could define this special string: "AcceptAll"; When "AcceptAll".equals(serverCACertFile) holds, we trust all certificate. If serverCACertFile is null, we throw an exception and hence users know HttpsSegmentFetcher is not working properly. After former two cases, we assume that serverCACertFile is the path to a file.